lib/std/debug.zig

```diff
@@ -1036,44 +1056,145 @@ pub fn readElfDebugInfo(allocator: mem.Allocator, elf_file: File) !ModuleDebugIn
-                opt_debug_str_offsets = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_line")) {
-                opt_debug_line = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_line_str")) {
-                opt_debug_line_str = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_ranges")) {
-                opt_debug_ranges = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_loclists")) {
-                opt_debug_loclists = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_rnglists")) {
-                opt_debug_rnglists = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_addr")) {
-                opt_debug_addr = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_names")) {
-                opt_debug_names = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
-            } else if (mem.eql(u8, name, ".debug_frame")) {
-                opt_debug_frame = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
+
+            if (mem.eql(u8, name, ".gnu_debuglink")) {
+                const gnu_debuglink = try chopSlice(mapped_mem, shdr.sh_offset, shdr.sh_size);
+                const debug_filename = mem.sliceTo(@ptrCast([*:0]const u8, gnu_debuglink.ptr), 0);
+                const crc_offset = mem.alignForward(@ptrToInt(&debug_filename[debug_filename.len]) + 1, 4) - @ptrToInt(gnu_debuglink.ptr);
+                const crc_bytes = gnu_debuglink[crc_offset .. crc_offset + 4];
+                separate_debug_crc = mem.readIntSliceNative(u32, crc_bytes);
+                separate_debug_filename = debug_filename;
+                continue;
+            }
+
+            const sections = [_]struct { name: []const u8, out: *?[]const u8 }{
+                .{ .name = ".debug_info", .out = &opt_debug_info },
+                .{ .name = ".debug_abbrev", .out = &opt_debug_abbrev },
+                .{ .name = ".debug_str", .out = &opt_debug_str },
+                .{ .name = ".debug_str_offsets", .out = &opt_debug_str_offsets },
+                .{ .name = ".debug_line", .out = &opt_debug_line },
+                .{ .name = ".debug_line_str", .out = &opt_debug_line_str },
+                .{ .name = ".debug_ranges", .out = &opt_debug_ranges },
+                .{ .name = ".debug_loclists", .out = &opt_debug_loclists },
+                .{ .name = ".debug_rnglists", .out = &opt_debug_rnglists },
+                .{ .name = ".debug_addr", .out = &opt_debug_addr },
+                .{ .name = ".debug_names", .out = &opt_debug_names },
+                .{ .name = ".debug_frame", .out = &opt_debug_frame },
```

Comment:
For stack unwinding you will most definitely want to include `.eh_frame` section. In fact, I haven't seen `.debug_frame` used much or at all in practice.

Reply:
> Very nice. I had no idea ELF had a debug symbols bundle-like mechanism in place too. Re stack unwinding using DWARF CFI perhaps this can be some inspiration [zig-dwarfdump/src/DwarfDump.zig#L928](https://github.com/kubkon/zig-dwarfdump/blob/01cf0b0af6a1fd3426f5be1a2ba781cb1395934f/src/DwarfDump.zig#L928).

Thanks, this looks like a great reference!

Depending on the amount of work involved (apparently the DWARF unwind stuff is Turing complete: http://kristerw.blogspot.com/2016/01/more-turing-completeness-in-surprising.html) I may move this PR into review before doing that, once I've cleaned things up a bit.

Reply:
Sounds good! Thanks for working on this!